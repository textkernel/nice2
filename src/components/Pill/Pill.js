import React from 'react';
import PropTypes from 'prop-types';
import PopupBase from '../PopupBase';
import PillButton from './PillButton';
import PillDropdown from './PillDropdown';

const Pill = props => {
    const {
        onClear,
        name,
        ref,
        content,
        children,
        dropdownRef: dropdownRefFromProps,
        ...rest
    } = props;

    const buttonRef = React.useMemo(() => ref || React.createRef(), [ref]);
    const dropdownRef = React.useMemo(() => dropdownRefFromProps || React.createRef(), [
        dropdownRefFromProps,
    ]);

    // eslint-disable-next-line react/display-name, react/prop-types
    const buttonRenderer = ({ setPopupVisibility, isOpen }) => {
        const toggleDropdown = () => {
            setPopupVisibility(!isOpen);
        };

        return (
            <PillButton
                name={name}
                content={content}
                isOpen={isOpen}
                toggleDropdown={toggleDropdown}
                onClear={onClear}
                {...rest}
            />
        );
    };

    // eslint-disable-next-line react/display-name, react/prop-types
    const dropdownRenderer = ({ setPopupVisibility }) => (
        <PillDropdown close={() => setPopupVisibility(false)} innerPadding="var(--spacing-3x)">
            {children}
        </PillDropdown>
    );

    return (
        <PopupBase
            anchorRenderer={buttonRenderer}
            popupRenderer={dropdownRenderer}
            anchorRef={buttonRef}
            popupRef={dropdownRef}
        />
    );
};

Pill.displayName = 'Pill';

Pill.propTypes = {
    /** The dropdown content renderer function. It is called with:
     *   * close {function} that closes the dropdown
     *   * innerPadding {string} that can be applied inside the component to set consistent padding
     */
    children: PropTypes.func.isRequired,
    /** a function to be called to clear the pill/filter content */
    onClear: PropTypes.func.isRequired,
    /** name describing the pill/filter */
    name: PropTypes.string.isRequired,
    /** label describing the content of an active filter/pill */
    content: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /** ref for pill button */
    ref: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    /** ref for pill dropdown */
    dropdownRef: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

Pill.defaultProps = {
    content: null,
    ref: null,
    dropdownRef: null,
};

export default Pill;
