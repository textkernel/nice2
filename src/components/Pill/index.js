/* eslint-disable import/prefer-default-export */

export { default as PillButton } from './PillButton';
// export { default as PillDropdown } from './PillDropdown';
// export {default as Pill } from './Pill'
