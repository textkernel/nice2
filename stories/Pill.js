import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import { boolean, text, withKnobs } from '@storybook/addon-knobs'; // eslint-disable-line import/no-extraneous-dependencies
import { PillButton } from '@textkernel/oneui';

storiesOf('Molecules|Pill', module)
    .addDecorator(withKnobs)
    .add('PillButton', () => (
        <PillButton
            isOpen={boolean('Pill is open', false)}
            togglePopup={() => {
                console.log('togglePopup has been called');
            }}
            onClear={() => {
                console.log('onClear has been called');
            }}
            name={text('Name of the pill', 'Pill name')}
            content={text('Label for pill content', 'This pill is used')}
        />
    ));
