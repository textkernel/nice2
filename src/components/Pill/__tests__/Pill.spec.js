import React from 'react';
import toJson from 'enzyme-to-json';
import Pill from '../Pill';

describe('<Pill> component', () => {
    const childrenMock = jest.fn();
    const onClearMock = jest.fn();
    const nameMock = 'Pill name';
    const contentMock = 'Pill content';

    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <Pill onClear={onClearMock} name={nameMock} content={contentMock} doneLabel="Done">
                {childrenMock}
            </Pill>
        );
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('PillButton')).toHaveLength(1);
        expect(wrapper.find('PillDropdown')).toHaveLength(0);
    });
    it('should open dropdown when button is clicked', () => {
        wrapper.find('.PillButton__pill').simulate('click');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('PillButton')).toHaveLength(1);
        expect(wrapper.find('PillDropdown')).toHaveLength(1);
    });
    it('should close dropdown when button is clicked again', () => {
        wrapper.find('.PillButton__pill').simulate('click');
        expect(wrapper.find('PillDropdown')).toHaveLength(1);

        wrapper.find('.PillButton__pill').simulate('click');
        expect(wrapper.find('PillDropdown')).toHaveLength(0);
    });
    it('should render children when dropdown is open', () => {
        expect(childrenMock).not.toHaveBeenCalled();

        wrapper.find('.PillButton__pill').simulate('click');
        expect(wrapper.find('PillDropdown')).toHaveLength(1);
        expect(childrenMock).toHaveBeenCalledTimes(1);
    });
});
