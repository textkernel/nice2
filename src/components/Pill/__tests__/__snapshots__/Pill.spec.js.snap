// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pill> component should open dropdown when button is clicked 1`] = `
<Pill
  additionalDropdownProps={Object {}}
  content="Pill content"
  doneLabel="Done"
  dropdownRef={null}
  isContentDefault={false}
  name="Pill name"
  noPaddingInDropdown={false}
  onClear={[MockFunction]}
  ref={null}
>
  <PopupBase
    anchorRef={
      Object {
        "current": <div
          class="PillButton"
        >
          <div
            class="PillButton__label PillButton__label--isActive"
          >
            Pill name
          </div>
          <div
            class="PillButton__pill PillButton__pill--isActive PillButton__pill--isOpen"
            role="button"
            tabindex="0"
          >
            <span>
              Pill content
            </span>
            <button
              class="PillButton__button PillButton__button--isActive PillButton__button--isOpen"
              type="button"
            >
              <svg
                class="PillButton__buttonIcon PillButton__buttonIcon--isActive PillButton__buttonIcon--isOpen"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                />
              </svg>
            </button>
          </div>
        </div>,
      }
    }
    anchorRenderer={[Function]}
    placement="bottom-start"
    popupRef={
      Object {
        "current": <div
          role="presentation"
        >
          <div
            class="PillDropdown"
            data-popup="true"
            role="dialog"
          >
            <div
              class="PillDropdown__content"
            />
            <div
              class="PillDropdown__footer"
            >
              <button
                class="Button Button--context_primary Button--size_small"
                type="button"
              >
                Done
              </button>
            </div>
          </div>
        </div>,
      }
    }
    popupRenderer={[Function]}
    renderInPortal={false}
  >
    <PillButton
      content="Pill content"
      isContentDefault={false}
      isOpen={true}
      name="Pill name"
      onClear={[MockFunction]}
      toggleDropdown={[Function]}
    >
      <div
        className="PillButton"
      >
        <div
          className="PillButton__label PillButton__label--isActive"
        >
          Pill name
        </div>
        <div
          className="PillButton__pill PillButton__pill--isActive PillButton__pill--isOpen"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
        >
          <span>
            Pill content
          </span>
          <button
            className="PillButton__button PillButton__button--isActive PillButton__button--isOpen"
            type="button"
          >
            <IoIosArrowDown
              className="PillButton__buttonIcon PillButton__buttonIcon--isActive PillButton__buttonIcon--isOpen"
            >
              <IconBase
                attr={
                  Object {
                    "viewBox": "0 0 512 512",
                  }
                }
                className="PillButton__buttonIcon PillButton__buttonIcon--isActive PillButton__buttonIcon--isOpen"
              >
                <svg
                  className="PillButton__buttonIcon PillButton__buttonIcon--isActive PillButton__buttonIcon--isOpen"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                    key="0"
                  />
                </svg>
              </IconBase>
            </IoIosArrowDown>
          </button>
        </div>
      </div>
    </PillButton>
    <PillDropdown
      close={[Function]}
      data-popup="true"
      doneLabel="Done"
      noPadding={false}
    >
      <div
        role="presentation"
      >
        <div
          className="PillDropdown"
          data-popup="true"
          role="dialog"
        >
          <div
            className="PillDropdown__content"
          />
          <div
            className="PillDropdown__footer"
          >
            <Button
              context="primary"
              disabled={false}
              href={null}
              isBlock={false}
              isInline={false}
              onClick={[Function]}
              size="small"
              type="button"
            >
              <button
                className="Button Button--context_primary Button--size_small"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Done
              </button>
            </Button>
          </div>
        </div>
      </div>
    </PillDropdown>
  </PopupBase>
</Pill>
`;

exports[`<Pill> component should render correctly 1`] = `
<Pill
  additionalDropdownProps={Object {}}
  content="Pill content"
  doneLabel="Done"
  dropdownRef={null}
  isContentDefault={false}
  name="Pill name"
  noPaddingInDropdown={false}
  onClear={[MockFunction]}
  ref={null}
>
  <PopupBase
    anchorRef={
      Object {
        "current": <div
          class="PillButton"
        >
          <div
            class="PillButton__label PillButton__label--isActive"
          >
            Pill name
          </div>
          <div
            class="PillButton__pill PillButton__pill--isActive"
            role="button"
            tabindex="0"
          >
            <span>
              Pill content
            </span>
            <button
              class="PillButton__button PillButton__button--isActive"
              type="button"
            >
              <span
                class="PillButton__buttonIcon PillButton__buttonIcon--isActive"
              >
                ×
              </span>
            </button>
          </div>
        </div>,
      }
    }
    anchorRenderer={[Function]}
    placement="bottom-start"
    popupRef={
      Object {
        "current": null,
      }
    }
    popupRenderer={[Function]}
    renderInPortal={false}
  >
    <PillButton
      content="Pill content"
      isContentDefault={false}
      isOpen={false}
      name="Pill name"
      onClear={[MockFunction]}
      toggleDropdown={[Function]}
    >
      <div
        className="PillButton"
      >
        <div
          className="PillButton__label PillButton__label--isActive"
        >
          Pill name
        </div>
        <div
          className="PillButton__pill PillButton__pill--isActive"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
        >
          <span>
            Pill content
          </span>
          <button
            className="PillButton__button PillButton__button--isActive"
            onClick={[Function]}
            onKeyDown={[Function]}
            type="button"
          >
            <span
              className="PillButton__buttonIcon PillButton__buttonIcon--isActive"
            >
              ×
            </span>
          </button>
        </div>
      </div>
    </PillButton>
  </PopupBase>
</Pill>
`;
